crates/iota-data-ingestion/src/workers/historical.rs

```diff
@@ -0,0 +1,142 @@
+
+use crate::RelayWorker;
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+#[serde(rename_all = "kebab-case")]
+pub struct HistoricalWriterConfig {
+    pub object_store_config: ObjectStoreConfig,
+    pub commit_duration_seconds: u64,
+}
+
+pub struct HistoricalReducer {
+    remote_store: Arc<DynObjectStore>,
+    commit_duration_ms: u64,
+}
+
+impl HistoricalReducer {
+    pub async fn new(config: HistoricalWriterConfig) -> anyhow::Result<Self> {
+        let remote_store = config.object_store_config.make()?;
+
+        Ok(Self {
+            remote_store,
+            commit_duration_ms: config.commit_duration_seconds * 1000,
+        })
+    }
+
+    async fn upload(
+        &self,
+        start: CheckpointSequenceNumber,
+        end: CheckpointSequenceNumber,
+        data: Bytes,
+    ) -> anyhow::Result<()> {
+        let file_metadata = create_file_metadata_from_bytes(data.clone(), start..end)?;
+        self.remote_store
+            .put(&file_metadata.file_path(), data.into())
+            .await?;
+        let mut manifest = Self::read_manifest(&self.remote_store).await?;
+        manifest.update(end, file_metadata);
+
+        let bytes = finalize_manifest(manifest)?;
+        self.remote_store
+            .put(&Manifest::file_path(), bytes.into())
+            .await?;
+        Ok(())
+    }
+
+    fn prepare_data_to_upload(&self, mut checkpoint_data: Vec<u8>) -> anyhow::Result<Bytes> {
+        let mut buffer = vec![0; MAGIC_BYTES];
+        BigEndian::write_u32(&mut buffer, CHECKPOINT_FILE_MAGIC);
+        buffer.push(StorageFormat::Blob.into());
+        buffer.push(FileCompression::Zstd.into());
+        buffer.append(&mut checkpoint_data);
+        let mut compressed_buffer = vec![];
+        let mut cursor = Cursor::new(buffer);
+        compress(&mut cursor, &mut compressed_buffer)?;
+        Ok(Bytes::from(compressed_buffer))
+    }
+
+    pub async fn get_watermark(&self) -> anyhow::Result<CheckpointSequenceNumber> {
+        let manifest = Self::read_manifest(&self.remote_store).await?;
+        Ok(manifest.next_checkpoint_seq_num())
+    }
+
+    async fn read_manifest(remote_store: &dyn ObjectStore) -> anyhow::Result<Manifest> {
+        Ok(match remote_store.get(&Manifest::file_path()).await {
+            Ok(resp) => read_manifest_from_bytes(resp.bytes().await?.to_vec())?,
+            Err(ObjectStoreError::NotFound { .. }) => Manifest::new(0),
+            Err(err) => Err(err)?,
+        })
+    }
+}
+
+#[async_trait]
+impl Reducer<RelayWorker> for HistoricalReducer {
+    async fn commit(&self, batch: Vec<Arc<CheckpointData>>) -> Result<(), anyhow::Error> {
+        if batch.is_empty() {
+            anyhow::bail!("commit batch can't be empty");
+        }
+        let mut buffer = vec![];
+        let first_checkpoint = &batch[0];
+        let start_checkpoint = first_checkpoint.checkpoint_summary.sequence_number;
+        let last_checkpoint = start_checkpoint + batch.len() as u64;
```

Comment:
This is id of checkpoint that is actually not included in the batch. (the last checkpoint in the batch is `start_checkpoint + batch.len() - 1` if I understand the code correctly).

Could we adjust the name appropriately?

Reply:
I opted in using a `Range` value instead for clearer semantics: https://github.com/iotaledger/iota/pull/5942/commits/86c22db0ac09008f70166d26bd3791d177beacda